qiita_db/artifact.py

```diff
@@ -269,107 +284,163 @@ def create(cls, filepaths, artifact_type, name=None, prep_template=None,
+                         command_parameters, visibility_id,
+                         artifact_type_id, submitted_to_vamps)
+                     VALUES (%s, %s, %s, %s, %s, %s, %s)
+                     RETURNING artifact_id"""
+            sql_args = [gen_timestamp, cmd_id, dtype_id,
+                        cmd_parameters, visibility_id, atype_id, False]
+            qdb.sql_connection.TRN.add(sql, sql_args)
+            a_id = qdb.sql_connection.TRN.execute_fetchlast()
+            # Associate the artifact with its filepaths
+            fp_ids = qdb.util.insert_filepaths(
+                fps, a_id, atype, "filepath",
+                move_files=mv_files, copy=(not mv_files))
+            sql = """INSERT INTO qiita.artifact_filepath
+                        (artifact_id, filepath_id)
+                     VALUES (%s, %s)"""
+            sql_args = [[a_id, fp_id] for fp_id in fp_ids]
+            qdb.sql_connection.TRN.add(sql, sql_args, many=True)
+            qdb.sql_connection.TRN.execute()
 
-        timestamp = datetime.now()
+            return cls(a_id)
 
-        with qdb.sql_connection.TRN:
-            visibility_id = qdb.util.convert_to_id("sandbox", "visibility")
-            artifact_type_id = qdb.util.convert_to_id(
-                artifact_type, "artifact_type")
+        def _associate_with_study(instance, study_id):
+            # Associate the artifact with the study
+            sql = """INSERT INTO qiita.study_artifact
+                        (study_id, artifact_id)
+                     VALUES (%s, %s)"""
+            sql_args = [study_id, instance.id]
+            qdb.sql_connection.TRN.add(sql, sql_args)
+            qdb.sql_connection.TRN.execute()
 
-            if parents:
-                # Check that all parents belong to the same study
-                studies = {p.study.id for p in parents}
-                if len(studies) > 1:
-                    raise qdb.exceptions.QiitaDBArtifactCreationError(
-                        "parents from multiple studies provided: %s"
-                        % ', '.join(studies))
-                study_id = studies.pop()
+        def _associate_with_analysis(instance, analysis_id):
+            # Associate the artifact with the analysis
+            sql = """INSERT INTO qiita.analysis_artifact
+                        (analysis_id, artifact_id)
+                     VALUES (%s, %s)"""
+            sql_args = [analysis_id, instance.id]
+            qdb.sql_connection.TRN.add(sql, sql_args)
+            qdb.sql_connection.TRN.execute()
 
-                # Check that all parents have the same data type
+        with qdb.sql_connection.TRN:
+            if parents:
                 dtypes = {p.data_type for p in parents}
-                if len(dtypes) > 1:
+                # If an artifact has parents, it can be either from the
+                # processing pipeline or the analysis pipeline. Decide which
+                # one here
+                studies = set()
```

Comment:
I actually decided to make a more realistic test and it turns out that it doesn't matter because the bulk of the time is spent on accessing the DB:
```python

In [33]: artifacts = [Artifact(randint(1, 9)) for i in range(1000)]

In [34]: %cpaste
Pasting code; enter '--' alone on the line to stop or use Ctrl-D.
:def f(parents):
:    studies = set()
:    analyses = set()
:    for p in parents:
:        s = p.study
:        a = p.analysis
:        if s is not None:
:            studies.add(s.id)
:        if a is not None:
:            analyses.add(a.id)
:    return studies, analyses
:
:def g(parents):
:    studies = {p.study for p in parents}
:    analyses = {p.analysis for p in parents}
:    studies.discard(None)
:    analyses.discard(None)
:    studies = {s.id for s in studies}
:    analyses = {a.id for a in analyses}
:    return studies, analyses
:--

In [35]: %timeit f(artifacts)
1 loops, best of 3: 1.25 s per loop

In [36]: %timeit g(artifacts)
1 loops, best of 3: 1.26 s per loop

```

Anyways I changed to the "g" version since it is more compact